### PR TITLE
CORE: MC/UC: reductions refactor

### PR DESCRIPTION
**Pull request by vspetrov**
## What
Moves all the reductions from MC to EC
Refactors EE task interface
Extends EE Task Reduce for multi-buffer support 
Unifies/Simplifies old ucc_mc_reduce, ucc_mc_reduce_multiple, ucc_mc_reduce_multiple_alpha into 1 new EE task ucc_ee_task_reduce_strided
Updates all the TL/UCP collectives to use executor (when required) instead of MC (blocking reductions are still blocking so far, will be changed in separate PR)

## Why ?
We needed a way to add new varieties of EE tasks (another iface for reduction for TL/SHM, extended COPY tasks, etc)
Separation of reduce functionality from MC (memory component) to EC

### CODE REVIEW

File: src/components/ec/cuda/kernel/ec_cuda_reduce.cu [hunk 1]
```diff
@@ -0,0 +1,336 @@
...
+            }                                                                  \
+            break;                                                             \
+        case 3:                                                                \
+            for (i = start; i < count; i += step) {                            \
+                d[i] = _OP##_4(s1[i], s2[i], s2[i + ld], s2[i + 2 * ld]);      \
+            }                                                                  \
+            break;                                                             \
+        default:                                                               \
+            for (i = start; i < count; i += step) {                            \
+                d[i] = _OP(s1[i], s2[i]);                                      \
```
Comment (samnordmann) [inline src/components/ec/cuda/kernel/ec_cuda_reduce.cu:96]: Same remark as for line 50

File: src/components/ec/cuda/kernel/ec_cuda_reduce.cu [hunk 2]
```diff
@@ -0,0 +1,336 @@
...
+            }                                                                  \
+            break;                                                             \
+        case 2:                                                                \
+            for (i = start; i < count; i += step) {                            \
+                d[i] = _OP##_3(s1[i], s2[i], s2[i + ld]);                      \
+            }                                                                  \
+            break;                                                             \
+        case 3:                                                                \
+            for (i = start; i < count; i += step) {                            \
+                d[i] = _OP##_4(s1[i], s2[i], s2[i + ld], s2[i + 2 * ld]);      \
```
Comment (samnordmann) [inline src/components/ec/cuda/kernel/ec_cuda_reduce.cu:91]: Would it be preferable to move the striding logic to ucc_dt_reduce_strided in ucc_dt_reduce.h (as proposed in a previous comment)? Maybe we would avoid duplicating CUDA_REDUCE_WITH_OP for the strided version

**Review by samnordmann — APPROVED** [review]

File: src/components/ec/base/ucc_ec_base.h [hunk 3]
```diff
@@ -63,10 +63,9 @@ enum ucc_ee_executor_params_field {
 };
 
 typedef enum ucc_ee_executor_task_type {
-    UCC_EE_EXECUTOR_TASK_TYPE_COPY               = UCC_BIT(0),
-    UCC_EE_EXECUTOR_TASK_TYPE_REDUCE             = UCC_BIT(1),
-    UCC_EE_EXECUTOR_TASK_TYPE_REDUCE_MULTI       = UCC_BIT(2),
-    UCC_EE_EXECUTOR_TASK_TYPE_REDUCE_MULTI_ALPHA = UCC_BIT(3),
+    UCC_EE_EXECUTOR_TASK_REDUCE         = UCC_BIT(0),
+    UCC_EE_EXECUTOR_TASK_REDUCE_STRIDED = UCC_BIT(1),
```
[the remark this answers is not in the document]
Comment (vspetrov) [reply src/components/ec/base/ucc_ec_base.h]: the reason we keep it separate is that with strided we can describe task with multiple vectors using just 3 values. Keeping task_args as small is possible is critical for cpu->gpu signalling. So, it is optimization.

File: src/components/ec/cpu/ec_cpu.c [hunk 4]
```diff
@@ -112,49 +112,54 @@ ucc_status_t ucc_cpu_executor_task_post(ucc_ee_executor_t *executor,
...
+    {
+        ucc_eee_task_reduce_strided_t *trs =
+            (ucc_eee_task_reduce_strided_t *)&task_args->reduce_strided;
+        size_t                n_srcs = trs->n_src2 + 1;
+        uint16_t              flags  = task_args->flags;
+        void **               srcs;
+        ucc_eee_task_reduce_t tr;
+        int                   i;
+
+        if (n_srcs <= UCC_EE_EXECUTOR_NUM_BUFS) {
```
[the remark this answers is not in the document]
Comment (vspetrov) [reply src/components/ec/cpu/ec_cpu.c:132]: see comment above: we want to have strided for the more optimized signalling in cuda executor

File: src/components/ec/cpu/ec_cpu_reduce.c [hunk 5]
```diff
@@ -0,0 +1,323 @@
...
+            break;                                                             \
+        default:                                                               \
+            ec_error(&ucc_ec_cpu.super,                                        \
+                     "float dtype does not support "                           \
+                     "requested reduce op: %s",                                \
+                     ucc_reduction_op_str(_op));                               \
+            return UCC_ERR_NOT_SUPPORTED;                                      \
+        }                                                                      \
+        if (flags & UCC_EEE_TASK_FLAG_REDUCE_WITH_ALPHA) {                     \
+            VEC_OP(d, _count, task->alpha);                                    \
```
[the remark this answers is not in the document]
Comment (vspetrov) [reply src/components/ec/cpu/ec_cpu_reduce.c:198]: Yes, there would be less "loads". But as you mentioned the code will be much more complex while we still not see much request for UCC_OP_AVG (so it is rare corner case).

File: src/components/ec/cpu/ec_cpu_reduce.c [hunk 6]
```diff
@@ -0,0 +1,323 @@
...
+            for (_j = 2; _j < _n_srcs; _j++) {                                 \
+                _tmp = _OP(_tmp, bfloat16tofloat32(&_s[_j][_i]));              \
+            }                                                                  \
+            float32tobfloat16(_tmp *_alpha, &_d[_i]);                          \
+        }                                                                      \
+    } while (0)
+
+#define DO_DT_REDUCE_BFLOAT16(_srcs, _dst, _op, _count, _n_srcs)               \
+    do {                                                                       \
+        float _a = (flags & UCC_EEE_TASK_FLAG_REDUCE_WITH_ALPHA) ? task->alpha \
```
[the remark this answers is not in the document]
Comment (vspetrov) [reply src/components/ec/cpu/ec_cpu_reduce.c:143]: with bfloat16 on CPU we already pay super costly bfloat16<->float conversion on each step. so, i think this multiplication is negligible. I can do the same "if" in the end though, to be similar to other dtypes.

File: src/components/ec/cuda/kernel/ec_cuda_reduce.cu (hunk 2 above)
Comment (vspetrov) [reply src/components/ec/cuda/kernel/ec_cuda_reduce.cu:91]: not sure what do you mean by "move striding logic to ucc_dt_reduce_striedd" ? DO you mean converting to default reduce ? Then, again, the answer is that we want to keep task_reduce_strided for better signalling in ee_cuda.

Comment (samnordmann) [reply src/components/ec/cuda/kernel/ec_cuda_reduce.cu:91]: yes, this is what I meant. Ok for the explanation

**Comment by vspetrov** [comment]
@edgargabriel plz have a look. This PR requires the corresponding changes for MC/EC/ROCM. You mentioned that you can prep a PR against my branch so that everything goes to upstream/master together

**Comment by edgargabriel** [comment]
@vspetrov I will work on it, I will try to have it ready by the end of the week

**Comment by edgargabriel** [comment]
> @edgargabriel plz have a look. This PR requires the corresponding changes for MC/EC/ROCM. You mentioned that you can prep a PR against my branch so that everything goes to upstream/master together

@vspetrov I filed the pr against your branch, please have a look at let me know. (https://github.com/vspetrov/ucc/pull/3). Not entirely sure why it wants to merger your commit as well as the new stuff as part of the pr (I did use your branch as the starting point of my branch), but we'll find a solution to resolve this, the code is fundamentally there and tested.

File: src/components/ec/base/ucc_ec_base.h [hunk 7]
```diff
@@ -77,6 +76,44 @@ typedef struct ucc_ee_executor_params {
...
+        void **srcs_ext;
+    };
+    size_t   count;
+    double   alpha;
+    uint16_t n_srcs;
+} ucc_eee_task_reduce_t;
+
+typedef struct ucc_eee_task_reduce_strided {
+    ucc_datatype_t     dt;
+    ucc_reduction_op_t op;
```
[the remark this answers is not in the document]
Comment (vspetrov) [reply src/components/ec/base/ucc_ec_base.h:95]: i've reorganized a bit reducing from 72 bytes to 64. Still the size of ucc_eee_task_args will be defined by the ucc_ee_task_copy_multi (under union) which is 152 currently. 

Still if we optimize signalling (pidx is int, if we split it into 16bits task type and 16 bits actual pidx) then we can read only sizeof(task_reduce_strided) inside executor kernel

**Comment by vspetrov** [comment]
@edgargabriel could you plz rebase your PR on top of latest changes in the branch. I've addressed comments and now it is incompatible with you changes. I assume you have changes in a separate PR so should be easy to rebase

**Comment by edgargabriel** [comment]
> @edgargabriel could you plz rebase your PR on top of latest changes in the branch. I've addressed comments and now it is incompatible with you changes. I assume you have changes in a separate PR so should be easy to rebase

@vspetrov done

File: src/components/ec/cuda/kernel/ec_cuda_reduce.cu [hunk 8]
```diff
@@ -0,0 +1,336 @@
...
+#if SIZEOF_CUDOUBLECOMPLEX == 16
+        DT_REDUCE_FLOAT_COMPLEX(cuDoubleComplex, double, task, op, stream, bk, th);
+        break;
+#else
+        return UCC_ERR_NOT_SUPPORTED;
+#endif
+#if CUDART_VERSION >= 11000
+    case UCC_DT_BFLOAT16:
+        ucc_assert(2 == sizeof(__nv_bfloat16));
+        DT_REDUCE_FLOAT(__nv_bfloat16, task, op, stream, bk, th);
```
[the remark this answers is not in the document]
Comment (vspetrov) [reply src/components/ec/cuda/kernel/ec_cuda_reduce.cu:322]: I checked. We don't need. nvcc will generate bf16<->f32 conversion for arch < 80.

**Review by Sergei-Lebedev — APPROVED** [review]

File: src/components/ec/cuda/kernel/ec_cuda_reduce.cu [hunk 9]
```diff
@@ -0,0 +1,340 @@
...
+        break;
+#endif
+    default:
+        ec_error(&ucc_ec_cuda.super, "unsupported reduction type (%s)",
+                 ucc_datatype_str(dt));
+        return UCC_ERR_NOT_SUPPORTED;
+    }
+    CUDA_CHECK(cudaGetLastError());
+    CUDA_CHECK(cudaStreamSynchronize(stream));
```
Comment (Sergei-Lebedev) [inline src/components/ec/cuda/kernel/ec_cuda_reduce.cu:335]: cudaStreamSynchronize is not needed, if executor is the only interface for reductions then ucc_executor_task_test should indicate about completion. Removing stream synchronize will make reduce async

Comment (vspetrov) [reply src/components/ec/cuda/kernel/ec_cuda_reduce.cu]: done

**Review by bureddy — APPROVED** [review]

File: src/components/ec/base/ucc_ec_base.h [hunk 10]
```diff
@@ -77,32 +76,83 @@ typedef struct ucc_ee_executor_params {
...
 
-typedef struct ucc_ee_executor_task_args_copy_multi{
+/* Reduces "n_srcs" buffers (each contains "count" elements of type "dt")
+   into "dst" buffer.
+
+   If UCC_EEE_TASK_FLAG_REDUCE_SRCS_EXT flag is not set on task_args
+   then the sources are taken from "srcs" and n_srcs can not exceed
+   UCC_EE_EXECUTOR_NUM_BUFS.
+
+   If UCC_EEE_TASK_FLAG_REDUCE_SRCS_EXT flag IS set on task_args
```
Comment (manjugv) [inline src/components/ec/base/ucc_ec_base.h:87]: IS -> is

File: src/components/ec/base/ucc_ec_base.h [hunk 11]
```diff
@@ -77,32 +76,83 @@ typedef struct ucc_ee_executor_params {
...
+   If UCC_EEE_TASK_FLAG_REDUCE_WITH_ALPHA flag is set on task_args
+   each element of the result of reduction is multiplied by "alpha" */
+typedef struct ucc_eee_task_reduce {
+    void *             dst;
+    union {
+        void * srcs[UCC_EE_EXECUTOR_NUM_BUFS];
+        void **srcs_ext;
+    };
+    size_t   count;
+    double   alpha;
```
Comment (manjugv) [inline src/components/ec/base/ucc_ec_base.h:101]: align

**Pull request merged**